File: web/src/utils/urls/post-url.ts
import memoize from "lodash/memoize";
import { Account, AccountType } from "src/models/account";
import { Post } from "src/models/post";

/**
 * @returns String `/jobs/[post_slug_without_id]_by_[poster_slug_without_id]_[post_id]`
 */

export const _getPostUrl = (
  post: Pick<Post, "slug" | "id">,
  poster: Pick<Account, "slug"> & AccountType
) => {
  const where = poster.type === "Company" ? "at" : "by";
  return `/jobs/${[post.slug]}_${where}_${poster.slug}_${post.id}`;
};

export const getPostUrl = memoize(_getPostUrl);

File: api/src/_utils/post_url.rs
use crate::{
  account::model::{AccountType, CompactAccount},
  post::model::CompactPost,
};

pub fn get_post_url(post: &CompactPost, poster: &CompactAccount) -> String {
  let where_is = match poster.r#type {
    AccountType::Company { .. } => "at",
    _ => "by",
  };
  format!(
    "/jobs/{}_{}_{}_{}",
    post.slug, where_is, poster.slug, post.id
  )
}
